pkg/peertls/peertls_test.go

```diff
@@ -137,159 +120,519 @@ func TestVerifyPeerCertChains(t *testing.T) {
+			unrelatedChain,
+			&ErrVerifyCAWhitelist,
+			nil,
+		},
+		{
+			"certificate revocation - single revocation ",
+			TLSExtConfig{Revocation: true},
+			revokedLeafChain,
+			nil,
+			nil,
+			nil,
+		},
+		// {
+		// 	"certificate revocation - multiple, serial revocations",
+		// 	TLSExtConfig{Revocation: true},
+		// 	func() []*x509.Certificate {
+		// 		rev := new(Revocation)
+		// 		err = rev.Unmarshal(revokedLeafChain[0].ExtraExtensions[0].Value)
+		// 		testpeertls.PrintJson(rev, "rev1")
+		// 		assert.NoError(t, err)
+		//
+		// 		time.Sleep(1 * time.Second)
+		// 		_, chain, err := revokeLeaf(revokedLeafKeys, revokedLeafChain)
+		// 		assert.NoError(t, err)
+		//
+		// 		err = rev.Unmarshal(chain[0].ExtraExtensions[0].Value)
+		// 		testpeertls.PrintJson(rev, "rev2")
+		// 		assert.NoError(t, err)
+		//
+		// 		// err = revDB.Put(revokedLeafChain, chain[0].ExtraExtensions[0])
+		// 		return chain
+		// 	}(),
+		// 	nil,
+		// 	nil,
+		// 	nil,
+		// },
+		// {
+		// 	"certificate revocation - error (older timestamp)",
+		// 	TLSExtConfig{Revocation: true},
+		// 	func() []*x509.Certificate {
+		// 		rev := new(Revocation)
+		// 		err = rev.Unmarshal(revokedLeafChain[0].ExtraExtensions[0].Value)
+		// 		testpeertls.PrintJson(rev, "rev1")
+		// 		assert.NoError(t, err)
+		//
+		// 		time.Sleep(1 * time.Second)
+		// 		_, chain, err := revokeLeaf(revokedLeafKeys, revokedLeafChain)
+		// 		assert.NoError(t, err)
+		//
+		// 		err = rev.Unmarshal(chain[0].ExtraExtensions[0].Value)
+		// 		testpeertls.PrintJson(rev, "rev2")
+		// 		assert.NoError(t, err)
+		//
+		// 		// err = revDB.Put(revokedLeafChain, chain[0].ExtraExtensions[0])
+		// 		return chain
+		// 	}(),
+		// 	nil,
+		// 	nil,
+		// 	nil,
+		// },
 	}
 
 	for _, c := range cases {
 		t.Run(c.testID, func(t *testing.T) {
-			exts := ParseExtensions(c.config, c.whitelist)
+			opts := ParseExtOptions{
+				CAWhitelist: c.whitelist,
+				RevDB:       revDB,
+			}
+
+			exts := ParseExtensions(c.config, opts)
 			assert.Equal(t, 1, len(exts))
-			for i, e := range exts {
-				ok, err := e.f(leafCert.ExtraExtensions[0], [][]*x509.Certificate{{leafCert, caCert, rootCert}})
-				assert.Equal(t, c.expected[i].err, err)
-				assert.Equal(t, c.expected[i].ok, ok)
+			for _, e := range exts {
+				err := e.f(c.certChain[0].ExtraExtensions[0], [][]*x509.Certificate{c.certChain})
+				if c.errClass != nil {
+					assert.True(t, c.errClass.Has(err))
+				}
+				if c.err != nil {
+					assert.NotNil(t, err)
+				}
+				if c.errClass == nil && c.err == nil {
+					assert.NoError(t, err)
+				}
 			}
 		})
 	}
 }
+
+// NB: keys are in the reverse order compared to certs (i.e. first key belongs to last cert)!
+func newCertChain(length int) (keys []crypto.PrivateKey, certs []*x509.Certificate, _ error) {
+	for i := 0; i < length; i++ {
+		key, err := NewKey()
+		if err != nil {
+			return nil, nil, err
+		}
+		keys = append(keys, key)
+
+		var template *x509.Certificate
+		if i == length-1 {
```

Comment:
just clarification question: why does the index/length determine when to create the CA template?